### ISSUE TITLE
Unable to download from auth-protected URLs

### ISSUE THREAD

**Issue by Gh0stByte** [Battleman/zoomdl#20]
Encountering this issue while attempting to download a lecture. These lectures are behind a custom zoom setup for my school (could be part of the issue) where the URL begins with https://*schoolname*.zoom.etc

The following error occurs while attempting a download with a password
```
C:\Users\Gh0st\Downloads\Portable>zoomdl.exe -u "https://*redacted*" -p "gH92b$ZG"
Traceback (most recent call last):
  File "zoom_dl\__main__.py", line 20, in <module>
  File "zoom_dl\__init__.py", line 18, in main
  File "zoom_dl\zoomdl.py", line 135, in download
UnboundLocalError: local variable 'meet_id' referenced before assignment
[6720] Failed to execute script __main__
```

Current Environment is
>Python 3.7.3 (v3.7.3:ef4ec6ed12, Mar 25 2019, 22:22:05) [MSC v.1916 64 bit (AMD64)] on win32

Also tried on WSL with no success
> Linux Gh0st-PC 4.19.84-microsoft-standard #1 SMP Wed Nov 13 11:44:37 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

```
Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "./zoomdl/__main__.py", line 20, in <module>
  File "./zoomdl/zoom_dl/__init__.py", line 18, in main
  File "./zoomdl/zoom_dl/zoomdl.py", line 135, in download
UnboundLocalError: local variable 'meet_id' referenced before assignment
```

This error occurs when attempting a download which does not require a password but does require you to be logged in (in both environments.)

```No video URL in meta, going bruteforce
Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "./zoomdl/__main__.py", line 20, in <module>
  File "./zoomdl/zoom_dl/__init__.py", line 18, in main
  File "./zoomdl/zoom_dl/zoomdl.py", line 142, in download
  File "./zoomdl/zoom_dl/zoomdl.py", line 80, in get_page_meta
AttributeError: 'NoneType' object has no attribute 'group'
```

Perhaps incorporate an argument to supply a cookie/auth string to enable downloads

**Comment by Battleman** [comment]
Custom auth methods are tricky, because they highly vary between users. The only possible solution for that is cookies; but I've already looked at it, and it's far from obvious. I'll try again, but if anyone knows better, I'll gladly accept a PR...

**Comment by Gh0stByte** [comment]
I'll snoop around and see what I can find

**Comment by Gh0stByte** [comment]
From my testing, it seems that the ```_zm_kms``` cookie is what verifies authentication. 

Adding the cookie with cookiejar allows the bypass of the login screen and goes directly to the password view for protected recordings or to the video page for unprotected recordings.

**Comment by Gh0stByte** [comment]
From my testing, it seems that the ```_zm_kms``` cookie is what verifies authentication. This single cookie is all that is needed.

Adding the cookie with cookiejar allowed the bypass of the login screen and goes directly to the password view for protected recordings or to the video page for unprotected recordings.  

```
import requests
import http.cookiejar
import http.cookies

url = 'https://****.zoom.com/***'
cookies =  {
"_zm_kms": "aw1_*****",
}
s=requests.session()
s.verify=False
requests.utils.add_dict_to_cookiejar(s.cookies, cookies)
s.headers.update({'User-Agent':'Mozilla/5.0 (Windows NT 6.3; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/53.0.2785.116 Safari/537.36','Content-Type':'application/x-www-form-urlencoded'})

def main():
    r=s.get(url)
    print(r.content)
    if('Sign In' in str(r.content)):
        print("Not unlocked")
    else:
        print("unlocked")
    
if __name__ == '__main__':
    main()

```

**Comment by Gh0stByte** [comment]
This seems to fix the authentication, however, parsing the output seems to be an issue
https://github.com/Gh0stByte/zoomdl/blob/c6f06a9f301e0b451b481dc98a52a37e6b5cef67/zoom_dl/zoomdl.py#L25-L26

Following error when attempting parse
```
Traceback (most recent call last):
  File "/usr/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "./zoomdl/__main__.py", line 20, in <module>
  File "./zoomdl/zoom_dl/__init__.py", line 18, in main
  File "./zoomdl/zoom_dl/zoomdl.py", line 160, in download
  File "./zoomdl/zoom_dl/zoomdl.py", line 81, in get_page_meta
  File "./zoomdl/yaml/__init__.py", line 142, in full_load
  File "./zoomdl/yaml/__init__.py", line 114, in load
  File "./zoomdl/yaml/constructor.py", line 49, in get_single_data
  File "./zoomdl/yaml/composer.py", line 39, in get_single_node
  File "./zoomdl/yaml/parser.py", line 98, in check_event
  File "./zoomdl/yaml/parser.py", line 174, in parse_document_start
yaml.parser.ParserError: expected '<document start>', but found '<scalar>'
  in "<unicode string>", line 45, column 2:
    };
```

**Comment by DemGiran** [comment]
@Gh0stByte did you figure it out?

**Comment by Gh0stByte** [comment]
> 
> 
> @Gh0stByte did you figure it out?

@DemGiran,

I started uni a week late and so I've been swamped with homework and haven't had a chance to explore this any further. However, from my testing as stated above, injecting the ```_zm_kms``` cookie/authstr allows you to bypass the login and scrape the page. However, the formatting must be different because the application crashes afterwards. 

Another issue which could arise is requiring auth + a password, since it seems like unique zoom handles may have slightly different html formats.

Perhaps using beautiful soup ( #11) to scrape the metadata directly from the js would allow dual-compatibility, but I'm not sure. I'll try to see what's up later on this week if I can.

**Comment by Battleman** [comment]
Just a thought: did you check with `-v 0` that the password is correctly received? If you're on Linux/*nix, the way you input your password will be incorrectly interpreted. See this section of [the readme](https://github.com/Battleman/zoomdl/blob/master/README.md#about-quotes-important).

**Comment by chkiss** [comment]
I've got the same issue. Password interpreted correctly when checked with "-v 0", but I was already using single quotes anyway.

**Comment by Battleman** [comment]
Maybe [browser_cookie3](https://github.com/borisbabic/browser_cookie3) could be a lead. Ask user to log in using firefox/chrome then copy the cookies? or open a custom window with custom cookies?
I'll investigate that. See if that works at all, or if the cookie is too strongly connected to the browser (e.g. user-agent checking)

**Comment by Gh0stByte** [comment]
As long as you can grab the ```_zm_kms``` cookie post-auth, you shouldn't have an issue. There is no user-agent checking from my testing.

**Comment by metal450** [comment]
Not sure if this should be a separate issue (seems related), but I've got a Zoom video behind a "registration" page (I enter my email, name, & captcha to view it), which I cannot download.  The script reports "[ERROR] Video not found in page. Is it login-protected?".

Are such videos not downloadable?

**Comment by Battleman** [comment]
For the moment, they aren't. ZoomDL only supports password-protected (with a shared password) videos. Login are more tricky, because they are not always uniform (in particular, they may redirect to third-party login methods). browser_cookie3 seems promising, but I don't have time rn to implement it. Any PR welcome. Otherwise I'll do it when my studies calm down.

**Comment by metal450** [comment]
> Login are more tricky, because they are not always uniform (in particular, they may redirect to third-party login methods).

Personally, I've only ever seen one type - which doesn't use a redirect - so I suppose it's the most common?  The login screen looks like below, & the moment you enter your info it shows the video (doesn't seem to redirect anywhere off-site):

![2020-09-29 10 08 57](https://user-images.githubusercontent.com/1877409/94590668-e5672400-023b-11eb-8707-66d2f84841e6.png)

> but I don't have time rn to implement it. Any PR welcome. Otherwise I'll do it when my studies calm down.

Fair enough - fingers crossed for some point in the future, then :)

**Comment by Battleman** [comment]
OK, so I've been playing around with browser_cookie3. It seems doable, but I encounter problems when installing a third-party library (lz4) to a local directory (to make the "compiled" version). So for the moment, this solution is dependent on https://github.com/python-lz4/python-lz4/issues/205. 

If someone has a recording that requires credentials like SSO, I can try to make something work with filling in the credentials. If there is no captcha...

**Comment by Zhenghao-Dai** [comment]
Maybe can you add another parameter-cookie into the program. Users can catch the fresh cookies

**Comment by Battleman** [comment]
Probably fixed in 2021.02.15_beta with #40, thanks to @Gh0stByte! Keeping this opened until it's tested and merged into master.
